### ISSUE TITLE
Error: Uncompression program ‘gzip’ not found

### ISSUE THREAD

**Issue by jcs090218** [ubolonton/emacs-tree-sitter#85]
I am getting this error on Windows while installing this package. Is there a way to avoid this? Maybe use default zip instead of gzip? 😕 (This is most likely to be an enhancement request)

```
tree-sitter-langs.el:51:1:Error: Uncompression program ‘gzip’ not found
```

Thanks for this awesome package!

**Comment by ubolonton** [comment]
This seems to be a common issue with Emacs on Windows: https://stackoverflow.com/questions/29233629/emacs-uncompression-program-gzip-not-found-when-trying-to-open-info-files. Does any of the answer in that link help?

**Comment by jcs090218** [comment]
I was hoping  to use something else. Maybe like `zip` instead of `gzip` if this error occurs?

**Comment by morganwillcock** [comment]
Windows doesn't have a zip command/binary by default either, so trying to call zip wouldn't be any different from trying to call gzip. I actually stumbled across this issue while looking for something else, but from what I can tell installing a current version of this package shouldn't download the compressed version of the pre-built libraries. Could you say how you are installing it and which version you are using?

Modern versions of Windows do come with a (slightly buggy) build of bsdtar and that could probably work, but when I tried installing this package from MELPA I didn't get the same error you are seeing.

**Comment by jcs090218** [comment]
> Windows doesn't have a zip command/binary by default either, so trying to call zip wouldn't be any different from trying to call gzip.

Yes, in batch you do [this](https://superuser.com/questions/110991/can-you-zip-a-file-from-the-command-prompt-using-only-windows-built-in-capabili). There are a bunch of answers online so I am not going to tell you how to `zip`/`unzip` on Windows 10.  Anyway, I think Windows 10 has the built-in `tar` executable.  Maybe that could be the alternative?

> I actually stumbled across this issue while looking for something else, but from what I can tell installing a current version of this package shouldn't download the compressed version of the pre-built libraries. Could you say how you are installing it and which version you are using?

Well, I opened this issue 7 days ago. So it happens in that timeline. I am using [melpa]() and not [melpa-stable](), so I assumed I am using the latest version of this package. My operating system is Windows 10 and I don't think the OS has the pre-built `gzip` executable in it.  Here is the result when I type `where gzip`.

```
INFO: Could not find files for the given pattern(s).
```

However, on another machine also with Windows 10. I have gzip installed using cygwin and the issue doesn't appear.

> Modern versions of Windows do come with a (slightly buggy) build of bsdtar and that could probably work,

Yeah, the tar should work.

> but when I tried installing this package from MELPA I didn't get the same error you are seeing.

Did you visit the buffer `*Compile-Log*`? It doesn't appear automatically. I just tried it and it still gives me the same error log.

**Comment by ubolonton** [comment]
> I actually stumbled across this issue while looking for something else, but from what I can tell installing a current version of this package shouldn't download the compressed version of the pre-built libraries.

The dynamic modules that `tsc` downloads are no longer compressed, but the grammar bundles for `tree-sitter-langs` still are, so this is a valid issue.

> Yes, in batch you do [this](https://superuser.com/questions/110991/can-you-zip-a-file-from-the-command-prompt-using-only-windows-built-in-capabili). There are a bunch of answers online so I am not going to tell you how to `zip`/`unzip` on Windows 10.
 
AFAICT, that's for `.zip` files, which are not the same as `.tar.gz` files. The former are **archives of compressed files**, while the latter are **compressed archives of files**.

> Anyway, I think Windows 10 has the built-in `tar` executable. Maybe that could be the alternative?

I'll check the built-in `tar` when I have access to a Windows machine, probably in around 3 weeks. Do note that, even if that program works, the fix could still take a while, as `tree-sitter-langs` relies on Emacs's built-in support for compressed files instead of dealing with them directly. `tree-sitter-langs` can provide a temporary workaround, but a proper fix would have to be within Emacs itself.

**Comment by morganwillcock** [comment]
> but a proper fix would have to be within Emacs itself

I had a quick look. The decompression is from tar-mode so the programs used are defined in `jka-compr-compression-info-list`. I tried to modify the compression for gz files to stream through the native tar version, but it seems it doesn't properly support the type of stream passthrough that you can do with non-Windows versions of bsdtar. In theory I think this should work as a replacement decompressor configuration:
```
tar -cf - @-
```
...but on the build that comes with Windows this doesn't write anything to stdout. So I don't think it is easy to make a change to fix this within the Emacs configuration or within Emacs, because Windows doesn't seem to have any direct replacement program that can just deal with gzip streams. In fact, even if the above did work it isn't ideal because it effectively builds a new tar file from the contents of the original instead of just extracting the original.

The binary distribution of Emacs for Windows does come with a bzip2 binary though, and decompressing with that works fine. So in theory I think if everything was distributed as tar.bz2 files then all platforms should work, or perhaps you can just download the bzip2 versions when installing on Windows.

Side note: the reason I didn't see the gzip error is because the dll failed to load and the installation aborted early - I'll open a separate issue for that.

**Comment by ubolonton** [comment]
This was fixed by #89.
